mms/protocol/otf_message_handler.py

```diff
@@ -0,0 +1,241 @@
+        length = self._retrieve_int(conn)
+        msg['modelPath'] = self._retrieve_buffer(conn=conn, length=length)
+        msg['batchSize'] = self._retrieve_int(conn=conn)
+        length = self._retrieve_int(conn=conn)
+        msg['handler'] = self._retrieve_buffer(conn, length)
+        gpu_id = self._retrieve_int(conn=conn)
+        if gpu_id >= 0:
+            msg['gpu'] = gpu_id
+
+        return "load", msg
+
+    def _retrieve_model_inputs(self, conn, msg):
+        end = False
+        while end is False:
+            model_input = dict()
+            length = self._retrieve_int(conn)
+            if length > 0:
+                model_input['name'] = self._retrieve_buffer(conn, length)
+            elif length == END_OF_LIST:
+                end = True
+                continue
+
+            length = self._retrieve_int(conn)
+
+            model_input['contentType'] = self._retrieve_buffer(conn, length)
+            length = self._retrieve_int(conn)
+            value = self._retrieve_buffer(conn, length)
+            model_input['value'] = value
+            msg.append(model_input)
+
+    def _retrieve_request_batch(self, conn, msg):
+        end = False
+        while end is False:
+            req_batch = dict()
+            length = self._retrieve_int(conn)
+
+            if length > 0:
+                req_batch['requestId'] = self._retrieve_buffer(conn, length)
+            elif length == END_OF_LIST:
+                end = True
+                continue
+
+            length = self._retrieve_int(conn)
+            req_batch['contentType'] = self._retrieve_buffer(conn, length)
+
+            length = self._retrieve_int(conn)
+            if length == START_OF_LIST:  # Beginning of list
+                req_batch['modelInputs'] = list()
+                self._retrieve_model_inputs(conn, req_batch['modelInputs'])
+
+            msg.append(req_batch)
+
+    def _retrieve_inference_msg(self, conn):
+        msg = dict()
+        length = self._retrieve_int(conn)
+        msg['modelName'] = self._retrieve_buffer(conn, length)
+        length = self._retrieve_int(conn)
+        if length == START_OF_LIST:
+            msg['requestBatch'] = list()
+            self._retrieve_request_batch(conn, msg['requestBatch'])
+
+        return "predict", msg
+
+    def retrieve_msg(self, conn):
+        """
+        Retrieve a message from the socket channel.
+
+        :param conn:
+        :return:
+        """
+        # Validate its beginning of a message
+        version = self._retrieve_double(conn)
+        if version != 1.0:
+            raise MMSError(Err.DECODE_FAILED, "Invalid message received")
+
+        cmd = self._retrieve_int(conn)
+
+        if cmd == 1:
+            cmd, msg = self._retrieve_load_msg(conn)
+        elif cmd == 2:
+            cmd, msg = self._retrieve_inference_msg(conn)
+        else:
+            return "unknown", "Wrong command"
+
+        eom = self._retrieve_buffer(conn, 2)
+
+        if eom != b'\r\n':
```

Comment:
We should not have \r\n in protocol anymore

Reply:
Sure. I can remove that.